### PR TITLE
[FEAT] Implement hybrid search functionality with BM25 integration

### PR DESCRIPTION
**Pull request by bhavnicksm**
## Description

Add support for hybrid search by adding BM25S index fitting and search. 

We first initialize the BM25S index in the `fit` fn, and then during retrieval, based on the value of `alpha` we decide to combine it with vector search. `alpha` is a common value deciding the strength of the respective scores, which affects how much weight each component receives. 

Given that it required major code for combining the values together, I created a separate function `_merge_results` with the process to normalize the scores for each of the components, weighted addition based on `alpha`, re-rank them by combined scores, trim based on `k` value, and then present. 

`alpha` by default is set to `1.0` which implies that hybrid search is disabled

Please let me know if anything else is needed. Happy to receive feedback. 😊

## Copilot Summary
This pull request introduces significant enhancements to the `Pipeline` class in the `korok/korok.py` file, focusing on enabling hybrid search by integrating BM25 with vector search. The changes include updates to the initialization, fitting, and querying processes, as well as the addition of utility methods for normalizing and merging search results.

Key changes include:

### Hybrid Search Integration:

* Added `bm25` and `alpha` parameters to the `Pipeline` class to support hybrid search. The `bm25` parameter allows for the inclusion of a BM25 index, while the `alpha` parameter controls the weighting between BM25 and vector search results.
* Implemented the `is_hybrid_available` class method to check for the availability of the BM25 module.
* Updated the `fit` method to initialize and index BM25 if hybrid search is enabled (`alpha < 1.0`).
* Modified the `query` method to perform BM25 search and merge results with vector search results when hybrid search is enabled.

### Utility Methods for Result Handling:

* Added `_normalize_scores` to normalize search scores between 0 and 1.
* Added `_split_vicinity_results` to separate vector search results into documents and scores.
* Added `_merge_results` to combine and rank results from vector search and BM25 search based on the provided `alpha` value.

### Dependency Management:

* Added a new dependency group `hybrid` in `pyproject.toml` to include the `bm25s` package required for hybrid search.

### CODE REVIEW

**Review by stephantul — CHANGES_REQUESTED** [review]
Hello @bhavnicksm, 

Thanks a lot for the feature, it looks great. I think the code could use some polishing here and there, I've left some comments sprinkled throughout.

As for the optional bm25s dependency: as in the other PR, I would just add it as a required import for now, we can then later decide how to guard against optional imports. I feel like also introducing that here would lead to too big of a PR.

File: korok/korok.py [hunk 1]
```diff
@@ -1,37 +1,58 @@
...
+from importlib.util import find_spec
+from typing import TYPE_CHECKING, Any, List, Tuple
 
+import numpy as np
 from model2vec import StaticModel
 from vicinity import Backend, Metric, Vicinity
 
 from korok.rerankers import CrossEncoderReranker
 
+if TYPE_CHECKING:
```
[the remark this answers is not in the document]
Comment (bhavnicksm) [reply korok/korok.py:12]: So, `TYPE_CHECKING` conditionally imports modules for the static analysis phases of checking types but doesn't impact runtime (especially in cases of importing torch or scipy, the import is really slow). The modules inside are not actually imported. 

It's useful in scenarios where you are not importing a library but your methods are to intake or output an object from a library. For example, when outputing a `torch.Tensor` type, you would not want to `import torch` fully.

File: korok/korok.py [hunk 2]
```diff
@@ -1,37 +1,58 @@
...
+        corpus: list[str] | None = None,
     ) -> None:
         """
         Initialize a Pipeline instance.
 
         :param encoder: The encoder used to encode the items.
         :param vicinity: The vicinity object used to find nearest neighbors.
         :param reranker: The reranker used to rerank the results (optional).
+        :param bm25: The bm25 index used for hybrid search (optional).   
+        :param alpha: The alpha value for the hybrid search (optional).
```
[the remark this answers is not in the document]
Comment (bhavnicksm) [reply korok/korok.py:32]: Yes! Sorry about that, missed it~ will update the docstring to include corpus.

File: korok/korok.py [hunk 3]
```diff
@@ -51,8 +73,86 @@ def fit(
...
         )
-        return cls(encoder=encoder, vicinity=vicinity, reranker=reranker)
 
+        # Add bm25s if alpha > 0
+        if cls.is_hybrid_available() and alpha < 1.0:
+            global bm25s
+            import bm25s
+            
+            bm25 = bm25s.BM25()
+            tokens = bm25s.tokenize(texts, stopwords="en")
```
[the remark this answers is not in the document]
Comment (bhavnicksm) [reply korok/korok.py:78]: I'm not sure how multi-lingual this is, but an alternative to the "en" is to pass a `List[str]` with all the stopwords, I can make it configurable inside the `.fit` method

File: korok/korok.py [hunk 4]
```diff
@@ -51,8 +73,86 @@ def fit(
             metric=Metric.COSINE,
             **kwargs,
         )
-        return cls(encoder=encoder, vicinity=vicinity, reranker=reranker)
 
+        # Add bm25s if alpha > 0
+        if cls.is_hybrid_available() and alpha < 1.0:
```
[the remark this answers is not in the document]
Comment (bhavnicksm) [reply korok/korok.py:86]: Sure, we can have `hybrid` bool value to enable or disable `bm25`.

As for the `alpha` value, what would you suggest we rename it to? Something like `hybrid_search_weight` or `combination_factor`? I'm not sure what we could call it...?

**Comment by bhavnicksm** [comment]
Hey @stephantul! 

I believe I have made the necessary changes; could you please have a look? 

Thanks! 😊

File: korok/korok.py [hunk 5]
```diff
@@ -51,8 +73,86 @@ def fit(
...
+        denom = np.max(scores, axis=-1).reshape(-1, 1) - np.min(scores, axis=-1).reshape(-1, 1)
+        numerator = scores - np.min(scores, axis=-1).reshape(-1, 1)
+        return numerator / denom
+    
+    def _split_vicinity_results(self, results: List[List[Tuple[str, float]]]) -> List[List[Tuple[str, float]]]:
+        """Split the results from vector search into two lists."""
+        vicinity_docs = []
+        vicinity_scores = []
+        for i in range(len(results)):
+            vicinity_docs.append([str(results[i][j][0]) for j in range(len(results[i]))])
```
[the remark this answers is not in the document]
Comment (bhavnicksm) [reply korok/korok.py]: Result are a List[List[Tuple[str, float]]], so to simplify this code I distributed it into two for loops. Does that work?

File: korok/korok.py [hunk 6]
```diff
@@ -51,7 +81,89 @@ def fit(
...
+                if doc not in scores_list[i]:
+                    scores_list[i][doc] = Document(text=doc, bm25_score=score)
+                else:
+                    scores_list[i][doc].bm25_score = score
+
+        # Combine the docs and scores into a list of tuples
+        results = []
+        for i in range(len(vicinity_results)):
+            # Get the combined scores
+            results.append([(key, value.combine_scores(self.alpha))
```
Comment (stephantul) [inline korok/korok.py:157]: This append is unnecessary. You append the result here, and then immediately index below. Here's something that could also work: e.g.:

```python
result = [(key, value.combine_scores(self.alpha)) ...]
result = sorted(result, key=lambda x: x[1], reverse=True)
results[i] = result[:k]
```

File: korok/korok.py [hunk 7]
```diff
@@ -51,7 +81,89 @@ def fit(
...
+
+    def _merge_results(
+        self,
+        vicinity_results: List[List[Tuple[str, np.float64]]],
+        bm25_results: Tuple[List[List[str]], np.ndarray],
+        k: int = 10,
+    ) -> List[List[tuple[str, float]]]:
+        """Merge the results from vector search and bm25 search."""
+        # Initialize the scores list
+        scores_list = [{} for i in range(len(vicinity_results))]
```
Comment (stephantul) [inline korok/korok.py:129]: You now have the document class, so you could turn this into a `defaultdict` with `Document` as a default. That way, you don't need to check below if a certain key exists.

Comment (stephantul) [reply korok/korok.py:129]: Looking at it again, I don't think there's a need to pre-populate this list, you can just append to it as you go along.

File: korok/korok.py [hunk 8]
```diff
@@ -51,7 +81,89 @@ def fit(
...
+        min_score = np.min(scores, axis=-1).reshape(-1, 1)
+        max_score = np.max(scores, axis=-1).reshape(-1, 1)
+
+        # Normalize the scores
+        denom = max_score - min_score
+        numerator = scores - min_score
+
+        return numerator / denom
+
+    def _split_vicinity_results(self, results: List[List[Tuple[str, float]]]) -> List[List[Tuple[str, float]]]:
```
Comment (stephantul) [inline korok/korok.py:110]: I think the return type is incorrect. Did mypy flag this?

**Review by stephantul — APPROVED** [review]
Thanks, just some more small changes and then I think we're good to go!

File: korok/korok.py (hunk 8 above)
Comment (bhavnicksm) [reply korok/korok.py:110]: Oh, yes, it didn't get flagged for some reason; I will fix this! 

Does mypy need to be run separately? I did run ruff check and ruff format on the files once...

File: korok/korok.py (hunk 6 above)
Comment (bhavnicksm) [reply korok/korok.py:156]: Hey @stephantul, 

In the code, since `results[i]` has not been assigned earlier for that index, assigning it directly results in an error.

```
```

File: korok/korok.py (hunk 7 above)
Comment (bhavnicksm) [reply korok/korok.py]: Yes! That's a great suggestion; I didn't think of that earlier~ Since we don't have to check for the document if it exists or not anymore, it saves a lot of time...

**Comment by bhavnicksm** [comment]
Hey @stephantul!

Thank you for the detailed feedback~ 😊 I made the required changes, and tested out the working as well. Could you have a look? 

Thanks!

**Comment by bhavnicksm** [comment]
Just added both `sentence-transformers` and `bm25s` to the defualt installs. Does this seem fine, @Pringled?

**Comment by Pringled** [comment]
@bhavnicksm Sure, I think removing them is fine (and better) at the moment, and we can think about how to set this up at a later point, thanks!

**Comment by bhavnicksm** [comment]
Done! (^-^)

**Comment by Pringled** [comment]
Awesome, merging now! 🚀
